File: wallet/mobile/android/app/src/main/java/io/fourtwenty/wallet/NativeNetworkConfig420.kt
package io.fourtwenty.wallet

import android.content.Context
import org.json.JSONObject

/**
 * Configuration-only native network registry for W11.4.
 * Endpoints are supplied by build/runtime configuration; no signing authority lives here.
 */
data class NativeNetwork420(val chainId: String, val rpcUrl: String)

class NativeNetworkConfig420 private constructor(
    val networks: Map<String, NativeNetwork420>,
) {
    fun endpoints(): Map<String, String> = networks.mapValues { it.value.rpcUrl }

    companion object {
        fun fromJson(json: String): NativeNetworkConfig420 {
            val root = JSONObject(json)
            val entries = root.getJSONArray("networks")
            val resolved = linkedMapOf<String, NativeNetwork420>()
            for (index in 0 until entries.length()) {
                val item = entries.getJSONObject(index)
                val chainId = normalizeChainId(item.getString("chainId"))
                val rpcUrl = item.getString("rpcUrl")
                require(rpcUrl.startsWith("https://", ignoreCase = true)) { "RPC endpoint must use HTTPS" }
                require(URL420.isValidHttpsEndpoint(rpcUrl)) { "RPC endpoint is malformed" }
                require(chainId !in resolved) { "duplicate chain id" }
                resolved[chainId] = NativeNetwork420(chainId, rpcUrl)
            }
            require(resolved.isNotEmpty()) { "at least one network is required" }
            return NativeNetworkConfig420(resolved)
        }

        fun fromAsset(context: Context, assetName: String = "wallet-networks.json"): NativeNetworkConfig420 {
            val json = context.assets.open(assetName).bufferedReader().use { it.readText() }
            return fromJson(json)
        }

        fun normalizeChainId(value: String): String {
            val trimmed = value.trim().lowercase()
            require(Regex("^0x[0-9a-f]+$").matches(trimmed)) { "chain id must be canonical hex" }
            return "0x" + trimmed.removePrefix("0x").trimStart('0').ifEmpty { "0" }
        }
    }
}

private object URL420 {
    fun isValidHttpsEndpoint(value: String): Boolean = try {
        val url = java.net.URL(value)
        url.protocol.equals("https", ignoreCase = true) && !url.host.isNullOrBlank() && url.userInfo == null
    } catch (_: Exception) {
        false
    }
}

File: wallet/mobile/ios/Wallet420/NativeNetworkConfig420.swift
import Foundation

struct NativeNetwork420: Equatable {
    let chainID: String
    let rpcURL: URL
}

/// Configuration-only native network registry for W11.4.
/// Network endpoints are supplied by app/runtime configuration and never carry signing authority.
struct NativeNetworkConfig420 {
    let networks: [String: NativeNetwork420]

    init(jsonData: Data) throws {
        let object = try JSONSerialization.jsonObject(with: jsonData)
        guard let root = object as? [String: Any], let entries = root["networks"] as? [[String: Any]] else {
            throw NativeNetworkConfig420Error.invalidConfig
        }
        var resolved: [String: NativeNetwork420] = [:]
        for entry in entries {
            guard let rawChainID = entry["chainId"] as? String,
                  let rawURL = entry["rpcUrl"] as? String
            else { throw NativeNetworkConfig420Error.invalidConfig }
            let chainID = try Self.normalizeChainID(rawChainID)
            guard let url = URL(string: rawURL),
                  url.scheme?.lowercased() == "https",
                  url.host?.isEmpty == false,
                  url.user == nil,
                  url.password == nil
            else { throw NativeNetworkConfig420Error.insecureEndpoint }
            guard resolved[chainID] == nil else { throw NativeNetworkConfig420Error.duplicateChain }
            resolved[chainID] = NativeNetwork420(chainID: chainID, rpcURL: url)
        }
        guard !resolved.isEmpty else { throw NativeNetworkConfig420Error.invalidConfig }
        self.networks = resolved
    }

    init(bundle: Bundle = .main, resource: String = "wallet-networks", extension ext: String = "json") throws {
        guard let url = bundle.url(forResource: resource, withExtension: ext) else {
            throw NativeNetworkConfig420Error.missingConfig
        }
        try self.init(jsonData: Data(contentsOf: url))
    }

    var endpoints: [String: String] {
        networks.mapValues { $0.rpcURL.absoluteString }
    }

    static func normalizeChainID(_ value: String) throws -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard trimmed.range(of: "^0x[0-9a-f]+$", options: .regularExpression) != nil else {
            throw NativeNetworkConfig420Error.invalidChainID
        }
        let hex = String(trimmed.dropFirst(2)).drop(while: { $0 == "0" })
        return "0x" + (hex.isEmpty ? "0" : String(hex))
    }
}

enum NativeNetworkConfig420Error: Error {
    case missingConfig
    case invalidConfig
    case invalidChainID
    case insecureEndpoint
    case duplicateChain
}
